File: API/app/schemas/solicitation.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.schemas.pagination import PaginatedResponse, PaginationParams


class SolicitationListParams(PaginationParams):
    status: str | None = None
    tipo: str | None = None
    cliente_id: int | None = None
    termo: str | None = None


class SolicitationClientBrief(BaseModel):
    clientes_id: int
    nome: str | None = None
    cpf_cnpj: str | None = None
    celular01: str | None = None
    telefone: str | None = None
    flag_whatsapp: bool | None = None

    model_config = ConfigDict(from_attributes=True)


class SolicitationContractDraft(BaseModel):
    contrato_referencia_id: int | None = None
    cliente_id: int | None = None
    usuario_id_vendedor: int | None = None
    regra_juros_id: int | None = None
    plano_id: int | None = None
    qtde_dias: int | None = None
    percent_juros: float | None = None
    valor_empretismo: float | None = None
    valor_parcela: float | None = None
    recorrencia: bool | None = None
    aluguel: bool | None = None
    cobranca_segunda: bool | None = None
    cobranca_terca: bool | None = None
    cobranca_quarta: bool | None = None
    cobranca_quinta: bool | None = None
    cobranca_sexta: bool | None = None
    cobranca_sabado: bool | None = None
    cobranca_domingo: bool | None = None
    cobranca_feriado: bool | None = None
    cobranca_mensal: bool | None = None
    cobranca_quinzenal: bool | None = None
    frequencia_pagamento: str | None = None
    numero_parcelas: int | None = None


class SolicitationClientDraft(BaseModel):
    nome: str | None = None
    cpf_cnpj: str | None = None
    telefone: str | None = None
    celular01: str | None = None
    flag_whatsapp: bool = True


class SolicitationRead(BaseModel):
    id: int
    cliente_id: int | None = None
    cliente_nome: str | None = None
    session_id: int | None = None
    nome_informado: str | None = None
    telefone: str | None = None
    cpf_cnpj: str | None = None
    valor_pretendido: float | None = None
    frequencia_pagamento: str | None = None
    numero_parcelas: int | None = None
    tipo: str
    status: str
    vendedor_id: int | None = None
    vendedor_nome: str | None = None
    valor_parcela: float | None = None
    taxa_juros: float | None = None
    contrato_id: int | None = None
    usuario_id_aprovou: int | None = None
    usuario_nome_aprovou: str | None = None
    datahora_solicitacao: datetime
    datahora_aprovacao: datetime | None = None
    observacao: str | None = None

    model_config = ConfigDict(from_attributes=True)


class SolicitationDetailRead(SolicitationRead):
    cliente_existente: SolicitationClientBrief | None = None
    cliente_sugerido: SolicitationClientDraft
    contrato_sugerido: SolicitationContractDraft


class SolicitationListResponse(PaginatedResponse[SolicitationRead]):
    model_config = ConfigDict()


class SolicitationPendingCountRead(BaseModel):
    pendentes: int


class SolicitationLinkClientRequest(BaseModel):
    cliente_id: int


class SolicitationCompleteContractRequest(BaseModel):
    contrato_id: int
    status: str = "APROVADO"

File: FRONTEND/src/models/solicitation.ts
export interface SolicitationClientBrief {
  clientes_id: number
  nome: string | null
  cpf_cnpj: string | null
  celular01: string | null
  telefone: string | null
  flag_whatsapp: boolean | null
}

export interface SolicitationClientDraft {
  nome: string | null
  cpf_cnpj: string | null
  telefone: string | null
  celular01: string | null
  flag_whatsapp: boolean
}

export interface SolicitationContractDraft {
  contrato_referencia_id: number | null
  cliente_id: number | null
  usuario_id_vendedor: number | null
  regra_juros_id: number | null
  plano_id: number | null
  qtde_dias: number | null
  percent_juros: number | null
  valor_empretismo: number | null
  valor_parcela: number | null
  recorrencia: boolean | null
  aluguel: boolean | null
  cobranca_segunda: boolean | null
  cobranca_terca: boolean | null
  cobranca_quarta: boolean | null
  cobranca_quinta: boolean | null
  cobranca_sexta: boolean | null
  cobranca_sabado: boolean | null
  cobranca_domingo: boolean | null
  cobranca_feriado: boolean | null
  cobranca_mensal: boolean | null
  cobranca_quinzenal: boolean | null
  frequencia_pagamento: string | null
  numero_parcelas: number | null
}

export interface Solicitation {
  id: number
  cliente_id: number | null
  cliente_nome: string | null
  session_id: number | null
  nome_informado: string | null
  telefone: string | null
  cpf_cnpj: string | null
  valor_pretendido: number | null
  frequencia_pagamento: string | null
  numero_parcelas: number | null
  tipo: string
  status: string
  vendedor_id: number | null
  vendedor_nome: string | null
  valor_parcela: number | null
  taxa_juros: number | null
  contrato_id: number | null
  usuario_id_aprovou: number | null
  usuario_nome_aprovou: string | null
  datahora_solicitacao: string
  datahora_aprovacao: string | null
  observacao: string | null
}

export interface SolicitationDetail extends Solicitation {
  cliente_existente: SolicitationClientBrief | null
  cliente_sugerido: SolicitationClientDraft
  contrato_sugerido: SolicitationContractDraft
}

export interface SolicitationListFilters {
  page: number
  page_size: number
  status?: string
  tipo?: string
  cliente_id?: number
  termo?: string
}

export interface SolicitationListResponse {
  items: readonly Solicitation[]
  total: number
  page: number
  page_size: number
}

export interface SolicitationPendingCount {
  pendentes: number
}
